okhttp/src/jvmMain/kotlin/okhttp3/AsyncDns.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2022 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package okhttp3
+
+import java.net.InetAddress
+import java.net.UnknownHostException
+import java.util.concurrent.CountDownLatch
+import okio.IOException
+
+/**
+ * An async domain name service that resolves IP addresses for host names.
+ *
+ * The main implementations will typically be implemented using specific Dns libraries such as
+ *  * Android DnsResolver
+ *  * OkHttp DnsOverHttps
+ *  * dnsjava Resolver
+ *
+ * Implementations of this interface must be safe for concurrent use.
+ *
+ * The interface may be extended by specific AsyncDns implementations with additional records types,
+ * so as SCVB/HTTPS records, which may not be generally applicable.
+ */
+interface AsyncDns {
+  /**
+   * Query dns records for `hostname`, in the order they are received.
+   */
+  fun query(hostname: String, callback: Callback)
+
+  /**
+   * Callback to receive results from the Dns Queries.
+   */
+  interface Callback {
```

Comment:
There are a lot of subtle considerations in this API design, ordering, repeated methods, empty results.